src/thor/expansion_action.cc

```diff
@@ -46,12 +46,49 @@ void writeExpansionProgress(Expansion* expansion,
   if (exp_props.count(Options_ExpansionProperties_cost))
     expansion->add_costs(static_cast<uint32_t>(cost));
   if (exp_props.count(Options_ExpansionProperties_edge_status))
-    expansion->add_edge_status(status);
+    if (dedupe) {
+      expansion->add_edge_status_flags(status_flags);
+    } else {
+      expansion->add_edge_status(status);
+    }
   if (exp_props.count(Options_ExpansionProperties_edge_id))
     expansion->add_edge_id(static_cast<uint32_t>(edgeid));
   if (exp_props.count(Options_ExpansionProperties_pred_edge_id))
     expansion->add_pred_edge_id(static_cast<uint32_t>(prev_edgeid));
 }
+
+struct expansion_properties_t {
+  baldr::GraphId prev_edgeid;
+  // mask describing what statuses the edge has seen
+  uint8_t stages_mask;
+  float duration;
+  std::vector<midgard::PointLL> shape;
+  uint32_t distance;
+  float cost;
+
+  expansion_properties_t() = default;
+  expansion_properties_t(baldr::GraphId prev_edgeid,
+                         uint8_t stages_mask,
+                         float duration,
+                         uint32_t distance,
+                         std::vector<midgard::PointLL> shape,
+                         float cost)
+      : prev_edgeid(prev_edgeid), stages_mask(stages_mask), duration(duration), distance(distance),
+        shape(std::move(shape)), cost(cost){};
+
+  // check if status is higher or same – as we will keep track of the latest one
+  static bool is_latest_status(uint8_t current, uint8_t candidate) {
+    uint8_t first_significant = current | candidate;
+    // connected(LSb) is the highest status, so check from right
+    for (uint8_t i = 0; i < Expansion_EdgeStatus_EdgeStatus_ARRAYSIZE; i++) {
+      uint8_t mask = 1 << i;
+      if (!(first_significant & mask))
+        continue;
+      return candidate & mask;
+    }
+    return false;
+  }
```

Comment:
its not obvious to me why we need the complication of remembering all the statuses an edge has seen using a mask. it is true that in bi/multidirectional search an edge can be reached and settled multiple times (once for each search direction) but we dont need to know the prior statuses we only need to know the most important (lowest value) one its ever seen. tell me what im missing here. cant this just be:

```suggestion
  // check if status is higher or same – as we will keep track of the latest one
  static bool is_latest_status(uint8_t current, uint8_t candidate) {
    return candidate < current;
  }
```

Reply:
Hm, yes, that would simplify a lot of things. I implemented it this way because it was the proposed solution in the original issue (https://github.com/valhalla/valhalla/issues/4601). To be honest, I assumed it had some use case I wasn't aware of but would be useful anyway. @nilsnolde, could you please shed some insight on this?

Reply:
yeah, @kevinkreiser is of course right. some things only come to you when it's in code I guess :) sorry for the extra hassle I caused!